sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala

```diff
@@ -167,11 +169,25 @@ object ScalaNativePluginInternal {
         .withCompilerConfig(nativeConfig)
 
     interceptBuildException {
-      await(sbtLogger) { implicit ec: ExecutionContext =>
-        implicit def scope: Scope = sharedScope
-        Build
-          .buildCached(config)
-          .map(_.toFile())
+      val profiler =
+        if (AsyncProfilerLoader.isSupported())
+          Some(AsyncProfilerLoader.load())
+        else None
+      try {
+        config.logger.info(s"[async-profiler] starting profiler: $profiler")
+        profiler.foreach(_.execute("start,event=cpu,interval=100000"))
+        await(sbtLogger) { implicit ec: ExecutionContext =>
+          implicit def scope: Scope = sharedScope
+          Build
+            .buildCached(config)
+            .map(_.toFile())
+        }
+      } finally {
+        profiler.foreach { p =>
+          config.logger.info("[async-profiler] stop profiler")
+          p.execute("stop")
+          p.execute("flamegraph,file=profile.html")
+        }
```

Comment:
By putting it in `ScalaNativePluginInternal` it would be included in all sbt-plugins. Because the `AsyncProfilerLoader` would not be available in other sbt builds, it would fail to start the sbt instance. 
Instead we can put it only in `MyScalaNativePlugin` by overriding `nativeLink` task there. What's more we can put it in some dedicated higher-order functions, allowing us to wrap any part of code we'd like to profile. 
So then it might look like: 
```scala
nativeLink := withAsyncProfiling(flamegraphFile="profile.html", interval=10000){
  nativeLink.value
 }
```

Reply:
Yes, `MyScalaNativePlugin` is only internal override of `ScalaNativePlugin` with implementation defined in `ScalaNativePluginInternal`. `ScalaNativePlugin` is the public part which should be limited in terms of library dependencies and ammount of custom tasks/complexity, becouse that's the part we would need to preserve source/binary compatibility in next versions. 

Sorry if I've introduced any confusion. My initial though was that adding the profiling in `SNPluginInternal` would not be a good idea for 2 reasons: would introduce additional library dependencies and might add redundant overhead. 
Because of that we'd probably want to limit when profiling can be enabled. For our internal builds we can all dependencies as we wish, so we can add the implementation to MyScalaNativePlugin, and that's where we should focus. 
However, there might be some cases when we'd like to profile some 3rd party project. For that we can have 2 solutions: 
 1. Introduce an sbt plugin which would depend on `ScalaNativePlugin` similary to `MyScalaNativePlugin` and would introduce profiling tasks. I think that currently that might be too much 
 2. A simpler solution would be to structure profiling integration in such a way, that we would be able to easily copy-paste the integration to other project for ad-hoc integration. From what you've presented we should be able to limit the required code to 1 or 2 functions/tasks. We only need to make sure to not touch any internals, and instead work on the public API of `ScalaNativePlugin` If documented well we should be able to easily able profiling in custom projects this way. I think that for now that would be the best solution.

Reply:
Thank you for the detailed explanation!
It seems we had a bit different views on that:

- I thought we'd like to add an integration with AsyncProfiler to public sbt-plugin so that all scala-native users can easily run profiling on any projects.
- While you'd like to focus on profiling scala-native running for our own projects

I agree that providing profiling feature to public scala-native plugin is too much (that's why I though it might be enough to running async-profiler on benchmark).
Let's focus on `MyScalaNativePlugin` for now 👍  I guess we can somehow find a good design to decouple the profiler with task implementation which is easy to imitate in other projects.